File: mobile_client/lib/shop/user_config_service.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:mobile_client/constants/endpoint_constants.dart';
import 'package:mobile_client/models/environment.dart';
import 'package:mobile_client/models/s3_url.dart';
import 'package:mobile_client/services/user_auth_service.dart';

class UserConfigService extends ChangeNotifier {
  final UserAuthService _userAuthService;
  final String _baseUrl = '${Environment.apiBaseUrl}/$usersEndpoint';

  UserConfigService(this._userAuthService);

  Future<void> changeAvatar(context, String id) async {
    try {
      final response = await http.put(
        Uri.parse(getAvatarPath(id)),
        headers: _userAuthService.headerAuthOptions,
      );

      if (response.statusCode ~/ 100 == 2) {
        final S3URL newAvatar = S3URL.fromJson(jsonDecode(response.body));
        if (_userAuthService.curUser != null) {
          _userAuthService.curUser?.avatar = newAvatar.url;
          _userAuthService.notifyUserHasChanged();
          notifyListeners();
        }
      } else {
        showSnackBar(context, "Échec de mise à jour de l'avatar ");
      }
    } catch (error) {
      showSnackBar(
          context, "Erreur en mettant à jour l'avatar: ${error.toString()}");
    }
  }

  getAvatarPath(String id) {
    return id.isNotEmpty
        ? "$_baseUrl/$avatarEndpoint/$id"
        : "$_baseUrl/$customAvatarEndpoint";
  }

  Future<void> changeDrawnAvatar(String id) async {
    throw UnimplementedError("Method not implemented. ID: $id");
  }

  Future<void> changeTheme(String id) async {
    throw UnimplementedError("Method not implemented. ID: $id");
  }

  void showSnackBar(BuildContext context, String message) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(message)),
    );
  }
}

File: desktop-client/src/app/account/services/user-config/user-config.service.ts
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { MatSnackBar } from '@angular/material/snack-bar';
import { SERVER_URL_API } from '@app/constants/server-url-and-api-constant';
import { UserAuthService } from '@app/shared/services/user-auth/user-auth.service';
import { AVATAR_ENDPOINT, CUSTOM_AVATAR_ENDPOINT, USERS_ENDPOINT } from '@common/constants/endpoint-constants';
import { S3Url } from '@common/interfaces/url';
import { firstValueFrom } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class UserConfigService {
    private readonly baseUrl: string = `${SERVER_URL_API}/${USERS_ENDPOINT}`;

    constructor(
        private readonly http: HttpClient,
        private readonly userAuthService: UserAuthService,
        private readonly snackBar: MatSnackBar,
    ) {}

    async changeAvatar(id: string) {
        try {
            const { url } = await firstValueFrom(this.http.put<S3Url>(this.getAvatarPath(id), {}, this.userAuthService.requestOptions()));
            if (this.userAuthService.curUser) {
                this.userAuthService.curUser.avatar = url || this.userAuthService.curUser.avatar;
            }
        } catch (response: unknown) {
            if (response instanceof HttpErrorResponse) this.snackBar.open(response.error.message, 'close');
        }
    }

    getAvatarPath(id: string) {
        return id ? `${this.baseUrl}/${AVATAR_ENDPOINT}/${id}` : `${this.baseUrl}/${CUSTOM_AVATAR_ENDPOINT}`;
    }

    async changeDrawnAvatar(id: string): Promise<void> {
        throw new Error(`Method not implemented.${id}`);
    }
}
